tests/auth0_tests/test_middleware.py

```
@@ -61,17 +80,119 @@ def test_stale_cache_valid_token(self, mock_cache, mock_refresh):
+    return override_settings(MIDDLEWARE_CLASSES=classes)(fun)
+
+
+class UserifiedClientHandler(ClientHandler):
+    """Enhances ClientHandler to "work" with users properly"""
+    def __init__(self, *args, **kwargs):
+        self.user = kwargs.pop('user')
+        super(UserifiedClientHandler, self).__init__(*args, **kwargs)
+
+    def get_response(self, req):
+        req.user = self.user
+        return super(UserifiedClientHandler, self).get_response(req)
+
+
+class ClientWithUser(Client):
+    """Enhances Client to "work" with users properly"""
+    def __init__(self, enforce_csrf_checks=False, **defaults):
+        # Start off with the AnonymousUser
+        self.user = AnonymousUser()
+        # Get this because we need to create a new UserifiedClientHandler later
+        self.enforce_csrf_checks = enforce_csrf_checks
+        super(ClientWithUser, self).__init__(**defaults)
+        # Stomp on the ClientHandler with one that correctly makes request.user
+        # the AnonymousUser
+        self.handler = UserifiedClientHandler(enforce_csrf_checks, user=self.user)
+
+    def login(self, **credentials):
+        from django.contrib.auth import authenticate
+
+        # Try to authenticate and throw an exception if that fails; also, this gets
+        # the user instance that was authenticated with
+        user = authenticate(**credentials)
+        if not user:
+            # Client lets you fail authentication without providing any helpful
+            # messages; we throw an exception because silent failure is
+            # unhelpful
+            raise Exception('Unable to authenticate with %r' % credentials)
+
+        ret = super(ClientWithUser, self).login(**credentials)
+        if not ret:
+            raise Exception('Login failed')
+
+        # Stash the user object it used and rebuild the UserifiedClientHandler
+        self.user = user
+        self.handler = UserifiedClientHandler(self.enforce_csrf_checks, user=self.user)
+        return ret
```

This was pretty rough. Here's what's going on.

Django's test client doesn't set `request.user` before merrily jaunting off into yonder field of daisies and Django's `RequestFactory` doesn't go through middleware at all.

Since we really want a few tests that test handling a request with the mechanics of the middleware to verify that the Django 1.10+ transitional mixin worked with our middleware correctly, so I was forced to do this weirdo thing where we set our own `ClientHandler` that adds a `.user` to the request before passing it through the machinery.

Another way to do this might be to switch to webtest, but I was too emotionally drained to see what shenanigans I would have to pull from my magic hat to get that vaguely working.

So that's why it looks so wonky. I tried to add lots of comments to alleviate the weirdness. Let me know if it's still unclear.

Also, I *hate* how Django's test client `.login` fails silently. That's so irritating and I lost at least an hour tinkering with code that was fine trying to debug it. In what bizarre and paranormal circumstances would I ever want it to fail silently, I cannot begin to fathom.